WordPress/Classes/ViewRelated/Blog/Site Picker/BlogList/BlogListSiteView.swift

```diff
@@ -45,40 +45,41 @@ private struct BlogListBadgeView: View {
     }
 }
 
-final class BlogListSiteViewModel: Identifiable {
-    var id: TaggedManagedObjectID<Blog> { TaggedManagedObjectID(blog) }
+struct BlogListSiteViewModel: Identifiable {
```

Comment:
I made it a reference because it has significantly less performance overhead when it's passed around or is stored in collections: no value copying and only a single ARC call per reference (as opposed to per-field ARC for structs). There are only disadvantages of making it a struct.

Reply:
struct is only copied on write, passing between functions or storing in collections should not come with any overhead. Plus, all properties in this type are immutable, so the struct should also not copied at runtime.

I profiled the app and couldn't see much of difference between using class or struct. I simply taped the ⬇️ button beside site title to show a site picker popover. Here is the result:

| `class` | `struct` |
| ------- | -------- |
| <img width="913" alt="Screenshot 2024-09-09 at 9 42 12 AM" src="https://github.com/user-attachments/assets/55ecfe9b-d7a5-4482-b7c1-6990b4fe2092"> | <img width="911" alt="Screenshot 2024-09-09 at 9 42 04 AM" src="https://github.com/user-attachments/assets/55c46b6b-c453-4f22-8b9d-e38a369e9ba5"> |

BTW, the memory modal of struct and class types don't seem to be an concern in [this documentation](https://developer.apple.com/documentation/swift/choosing-between-structures-and-classes). It's probably not something we need to be conscious of?

Reply:
> struct is only copied on write

It only applies to structs explicitly implementing Copy on Write, like `Array`. Regular structs are copied each time they are passed around, and ARC invokes retain/release for every property of a struct with reference types (e.g. `String` has one internally pointing to its buffer). Structs also often need to be moved to the heap, e.g. when they need to be put in an array etc, which requires an allocation anyway.

> I profiled the app and couldn't see much of difference between

You likely won't see much difference if you zoom one that much. It also depends on how these are being used. In this case, it won't make much different. But overall, structs generally have worse performance characteristics than reference types, which is why Copy of Write exists. The best case scenario is when structs are small and contain no reference types, like `Int` or `CGSize`.

There was a great, albeit long [session](https://developer.apple.com/videos/play/wwdc2016/416/) about it in 2016, but some parts might be outdated.


<img width="350" alt="Screenshot 2024-09-08 at 7 27 26 PM" src="https://github.com/user-attachments/assets/6aff2551-09a8-497b-906c-44e150a33aef">

<img width="350" alt="Screenshot 2024-09-08 at 7 27 29 PM" src="https://github.com/user-attachments/assets/dff65d0f-9417-4c79-b467-5f59947847e0">

Reply:
I'm aware of these technical differences between reference types and value types, their pros and cons, etc. What I'm not aware of are the optimizations that Swift compiler or LLVM may apply to a real code.

For example, in theory, class type's heap allocation has a higher performance cost than struct type, but I'm not sure if that's a legit concern unless we can profile the code that uses the type to show evidence.

I certainly hope that Apple would include more examples of when we should _not_ [Choose Structures by Default](https://developer.apple.com/documentation/swift/choosing-between-structures-and-classes#Choose-Structures-by-Default) in [the documentation](https://developer.apple.com/documentation/swift/choosing-between-structures-and-classes) I linked above.

> You likely won't see much difference if you zoom one that much. It also depends on how these are being used. In this case, it won't make much different.

I wanted to see how using class vs struct on this particular type would affect the app. And, since your concerns were creating many instances, storing them in collections, and passing them around functions, I have to use a wider range of interval to capture all those operations.

Do you mind showing a code snippet or unit test to show case your concerns on how using `BlogListSiteViewModel` as a struct may have a negative affect on the app's performance?

Reply:
> Do you mind showing a code snippet or unit test to show case your concerns on how using BlogListSiteViewModel as a struct may have a negative affect on the app's performance?

I doubt there is any significant difference, but going to the original question – what are the advantages of changing it to a struct? I'm not strongly opposed to changing it, I'm just wondering why change it?

As a rule of thumb, I think ViewModels should be classes especially if they have a non-trivial number of fields and/or some interaction, like in this case. I went with a class, for instance, to avoid putting copies into these arrays:

```swift
    @Published private(set) var recentSites: [BlogListSiteViewModel] = []
    @Published private(set) var allSites: [BlogListSiteViewModel] = []
    @Published private(set) var searchResults: [BlogListSiteViewModel] = []
```

Reply:
> what are the advantages of changing it to a struct?

I don't think there are really any particular _advantages_. At least there aren't anything that makes me think that we _definitely_ should choose struct (or class) for this `BlogListSiteViewModel` type.

> I'm just wondering why change it?

I don't really know how to explain my thinking other than it's just my intuition: I'd use a struct in the first place. It's a type that contains a few simple enough properties. All of those properties are immutable. struct seems to be a good choice for it.

> As a rule of thumb, I think ViewModels should be classes especially if they have a non-trivial number of fields and/or some interaction, like in this case.

I wouldn't declare `BlogListViewModel` as a struct. But I don't see why "Choose Structure by Default" principle can't be applied for `BlogListSiteViewModel`.

>  I went with a class, for instance, to avoid putting copies into these arrays:

Realistically, there are probably 10 or 20 instances in total stored in these arrays. A8C accounts are on another scale, which probably would have 900 instances.

> In terms of the performance, even Swift Core [team members](https://forums.swift.org/t/why-does-apple-recommend-to-use-structs-by-default/62066/5) confirmed that "big structs" are a legitimate issue.

I don't consider `BlogListSiteViewModel` to be "giant structs" though.

> As far as the "giant structs" problem goes, that's a legitimate issue with using value types in Swift today.

------

I feel like it's probably a matter of preference at this point. As I mentioned above, it's my intuition/preference to make `BlogListSiteViewModel` a value type. But I'm happy to put my preference aside and close this PR.